c_src/eleveldb.cc

```diff
@@ -703,40 +717,67 @@ async_write(
-    ERL_NIF_TERM result = fold(env, argv[2], write_batch_item, *batch);
-    if(eleveldb::ATOM_OK != result)
-    {
-        // must manually delete batch on failure at this point,
-        //  later WriteTask object will own and delete
-        delete batch;
-        return send_reply(env, caller_ref,
-                          enif_make_tuple3(env, eleveldb::ATOM_ERROR, caller_ref,
-                                           enif_make_tuple2(env, eleveldb::ATOM_BAD_WRITE_ACTION,
-                                                            result)));
+        // Seed the batch's data:
+        ERL_NIF_TERM result = fold(env, argv[2], write_batch_item, batch);
+        if(eleveldb::ATOM_OK != result)
+        {
+            return enif_make_tuple3(env, eleveldb::ATOM_ERROR, caller_ref,
+                                    enif_make_tuple2(env, eleveldb::ATOM_BAD_WRITE_ACTION,
+                                                     result));
+        }   // if
+
+        leveldb::Status status = db_ptr->m_Db->Write(opts, &batch);
+
+        return (status.ok() ? ATOM_OK : error_tuple(env, ATOM_ERROR_DB_WRITE, status));
     }   // if
+    else
+    {
+        // use heap if sending message :-(
+        leveldb::gPerfCounters->Inc(leveldb::ePerfDebug3);
 
-    leveldb::WriteOptions* opts = new leveldb::WriteOptions;
-    fold(env, argv[3], parse_write_option, *opts);
+        // Construct a write batch:
+        leveldb::WriteBatch* batch = new leveldb::WriteBatch;
 
-    eleveldb::WorkTask* work_item = new eleveldb::WriteTask(env, caller_ref,
-                                                            db_ptr.get(), batch, opts);
+        // Seed the batch's data:
+        ERL_NIF_TERM result = fold(env, argv[2], write_batch_item, *batch);
+        if(eleveldb::ATOM_OK != result)
+        {
+            delete batch;
+            return enif_make_tuple3(env, eleveldb::ATOM_ERROR, caller_ref,
+                                    enif_make_tuple2(env, eleveldb::ATOM_BAD_WRITE_ACTION,
+                                                     result));
+        }   // if
 
-    if(false == priv.thread_pool.Submit(work_item))
-    {
-        // work_item contains "batch" and the delete below gets both memory allocations
-        delete work_item;
-        return send_reply(env, caller_ref,
-                          enif_make_tuple2(env, eleveldb::ATOM_ERROR, caller_ref));
-    }   // if
+        leveldb::WriteOptions* opts = new leveldb::WriteOptions;
+        fold(env, argv[3], parse_write_option, *opts);
+
+        eleveldb::WorkTask* work_item = new eleveldb::WriteTask(env, caller_ref,
+                                                                db_ptr.get(), batch, opts);
+        eleveldb_priv_data& data = *static_cast<eleveldb_priv_data *>(enif_priv_data(env));
+        if(false == data.thread_pool.Submit(work_item))
+        {
+            // work_item contains "batch" and the delete below gets both memory allocations
```

Comment:
work_item also takes ownership of the opts variable that was allocated on the heap; it's probably worth updating the comment so that the next person in this code doesn't have to hunt down that fact